### PR TITLE
[autodiff] Handle multiple, mixed Independent Blocks (IBs) within multi-levels serial for-loops

### PR DESCRIPTION
**Pull request by erizmr**
Related issue = #3003
Current autodiff only handles multiple IBs between outermost loops (parallel loops). 
This PR aims to support handling **multiple**, **mixed** IBs within **multi-levels** serial for-loops.

| Cases  | Before  | After |
| ------------- | ------------- | ------------- |
| 1 | ❌   | ✅  |
| 2 | ✅ | ✅ |
| 3 | ✅ | ✅  |
| 4 | ✅  | ✅  |
| 5 | ❌ | ✅  |
| 6 | ✅  | ✅  |
| 7 | ❌  | ✅  |

Here are some examples:

- Multiple IBs with same nested for-loop (Multiple)
```py
for j in range(2):
    for i in range(3):
        # IB-1
        y[None] += x[None]
    for i in range(3):
        # IB-2
        y[None] += x[None]
```

- Multiple IBs with multi-level nested for-loop (Multiple + Multi-level)
```py
for j in range(2):
    for i in range(3):
        # IB-1
        y[None] += x[None]
    for i in range(3):
        for ii in range(2):
            # IB-2
            y[None] += x[None]
    for i in range(3):
        for ii in range(2):
            for iii in range(2):
                # IB-3
                y[None] += x[None]
```

- Multiple mixed IB/non-IBs with multi-level nested for-loop (Multiple + Mixed + Multi-level)
```py
for j in range(2):
    for i in range(3):
        # IB-1
        y[None] += x[None]
    for i in range(3):
        # IB-2
        for ii in range(2):
            y[None] += x[None]
        # Because the following block is non-ib, the IB falls back to parent level
        for iii in range(2):
            y[None] += x[None]
            for iiii in range(2):
                y[None] += x[None]
    for i in range(3):
        for ii in range(2):
            for iii in range(2):
                # IB-3
                y[None] += x[None]
```

### CODE REVIEW

File: tests/python/test_ad_for.py [hunk 1]
```diff
@@ -775,3 +775,159 @@ def test_large_loop():
...
+    x[None] = 1.0
+    with ti.Tape(y):
+        compute_y()
+
+    assert y[None] == 30.0
+    assert x.grad[None] == 30.0
+
+
+@test_utils.test(require=ti.extension.adstack)
+def test_multiple_ib_deeper():
```
Comment (ailzhang) [inline tests/python/test_ad_for.py:881]: It'd be great to add a test case that the input field isn't a scalar, just to make sure our test coverage is better ;)

File: taichi/transforms/auto_diff.cpp [hunk 2]
```diff
@@ -78,6 +81,51 @@ class IndependentBlocksJudger : public BasicStmtVisitor {
   bool is_inside_loop_ = false;
 };
 
+// Remove the duplicates IBs, remove blocks who are others' children because
```
Comment (ailzhang) [inline taichi/transforms/auto_diff.cpp:84]: nit: duplicated

**Review by ailzhang — APPROVED** [review]
LGTM!

File: taichi/transforms/auto_diff.cpp (hunk 2 above)
Comment (erizmr) [reply taichi/transforms/auto_diff.cpp]: Good catch. LOL

**Pull request merged**